vernac/comPrimitive.ml

```diff
@@ -38,7 +38,13 @@ let do_primitive id udecl prim typopt =
       Constrintern.(interp_type_evars_impls ~impls:empty_internalization_env)
         env evd typ
     in
-    let evd = Evarconv.unify_delay env evd typ expected_typ in
+    let evd = try Evarconv.unify_delay env evd typ expected_typ
+      with Evarconv.UnableToUnify (evd,e) as exn ->
+        let _, info = Exninfo.capture exn in
+        Exninfo.iraise (Pretype_errors.(
+            PretypeError (env,evd,CannotUnify (typ,expected_typ,Some e)),info))
```

Comment:
Maybe we should have a function implementing this unify-with-user-printable-error somewhere.
Incidentally a similar thing happens at https://github.com/coq/coq/blob/577ec77f17a872d6bc36073ceeb3cf582fcf01c4/vernac/comInductive.ml#L166-L169 but with `user_err (Himsg.explain_...)` instead of raising a PretypeError, not sure why.